### PR TITLE
Polish focus styles

### LINKED ISSUE #3039: Adjust and redesign neutral, hover, active and focus states
Given recent developments in design (see https://github.com/WordPress/gutenberg/issues/2983, https://github.com/WordPress/gutenberg/issues/2984, https://github.com/WordPress/gutenberg/issues/2980), it seems like now would be a good time to revisit how our buttons and controls could be unified and simplified in this context.

As such, this ticket builds on the past work in #127. It also subsumes #904. 

Challenge:

We have some new UI components and widgets in Gutenberg, compared to the rest of WordPress. These need to be accessible (see #127 again for lots of context), and have distinct focus styles.

Proposed design:

buttons focus states v2

Please open the above screenshot in a new tab to see the pixels. 

- All clickable items get a gray, button-shaped background color
- Active and focus states are the same

By using background shapes instead of borders, we can unify the styles across all the widgets we use in Gutenberg toolbars and menus.

Penny for your thoughts?

### LINKED ISSUE #3382: Improve menus styling consistency
At the moment, some dropdown menus have minor visual inconsistencies. For example: the block switcher menu and the Edit table menu have different padding around the focused menu item:

screen shot 2017-11-08 at 09 57 34

Also the button that opens the menus has a different "active" styling.

### PR DESCRIPTION
**Pull request by jasmussen**
This is a rather mammoth PR to improve focus styles across the board. Outside of the switch styles, which need more work and should be addressed separtely, this fixes #3382 and fixes #3039.

The impetus for this design is that we use a lot of "naked" components. Movers, ellipsis menus, icon-only buttons, which when given the traditional focus style appear immaterial. By giving a considered style and background, it gives solidity to the button.

Because it's a mammoth PR that touches everything at once, I've been able to reduce and trim a ton of styles, move many of the focus styles into mixins for reusability, and I've overall been able to do a ton of consistency work using SCSS variables. Border radii, menu item paddings, focus styles and alignments should all improve as of this. 

On the downside, we should test this PR pretty thoroughly.

**Edit**: Forgot to mention — this PR also both unifies and changes the _hover_ style. It's now _very dark gray_ everywhere, except for formatting controls where there's still that 1px border. 

I like this, I think it works, but there's not a lot of visual contrast between neutral and hover states now. I don't think it's a problem, but I wanted to surface it in case others do.

## Screenshots:

screen shot 2017-11-09 at 11 38 22
screen shot 2017-11-09 at 11 38 28
screen shot 2017-11-09 at 11 38 32
screen shot 2017-11-09 at 11 38 38
screen shot 2017-11-09 at 11 38 42
screen shot 2017-11-09 at 11 38 46
screen shot 2017-11-09 at 11 38 54
screen shot 2017-11-09 at 11 38 58
screen shot 2017-11-09 at 11 39 02
screen shot 2017-11-09 at 11 39 07
screen shot 2017-11-09 at 11 39 14
screen shot 2017-11-09 at 11 39 21

### CODE REVIEW

**Comment by youknowriad** [comment]
Awesome work. It's working pretty well globally.

Two questions:

 - The in-between blocks inserter has a different style, maybe you forgot about it?
 - Do we really need the "double" border on focus styles: the inner white and the black border. I think it may look better if we drop the white. Personal opinion though. (Maybe aside when we have both toggled and focus style at the same time)

File: editor/block-switcher/index.js [hunk 1]
```diff
@@ -110,9 +110,7 @@ function BlockSwitcher( { blocks, onTransform } ) {
 								} }
 								className="editor-block-switcher__menu-item"
 								icon={ (
-									<span className="editor-block-switcher__block-icon">
```
Comment (youknowriad) [inline editor/block-switcher/index.js]: The block icon is not always a dashicon, this wrapper was necessary to correctly add the spacing between the icon and text when we have custom elements as icons. Not saying we shouldn't remove it but it was there on purpose and we should make sure it doesn't break. #3238 

I think you're targetting `svg` but block icons are not always `svg`, they can be anything.

**Comment by jasmussen** [comment]
> The in-between blocks inserter has a different style, maybe you forgot about it?

I did, thanks!

> Do we really need the "double" border on focus styles: the inner white and the black border. I think it may look better if we drop the white. Personal opinion though. (Maybe aside when we have both toggled and focus style at the same time)

I think we do, as there are apparently some very specific requirements around contrast between _shapes_ and backgrounds, and the background itself doesn't count as a shape. 

But since we are moving this all to mixins, adjusting these in the future if we find that we can, should be easier to do in one place instead of 20 different scss files.

I restored the span and fixed things with a different approach. Good catch, Riad, thank you!

**Comment by jasmussen** [comment]
Rebased. I think this is ready for a final pair of eyes. CC: @mtias @karmatosed

**Comment by jasmussen** [comment]
Got a thumbs up in chat. Merging.

**Pull request merged**